src/plugins/streaming-ssr.js

```diff
@@ -0,0 +1,167 @@
+/** Copyright (c) 2018 Uber Technologies, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {createPlugin} from '../create-plugin';
+import {escape, consumeSanitizedHTML} from '../sanitization';
+import type {Context, SSRDecider as SSRDeciderService} from '../types.js';
+import string_stream from 'string-to-stream';
+import multi_stream from 'multistream';
+
+const SSRDecider = createPlugin({
+  provides: () => {
+    return ctx => {
+      // If the request has one of these extensions, we assume it's not something that requires server-side rendering of virtual dom
+      // TODO(#46): this check should probably look at the asset manifest to ensure asset 404s are handled correctly
+      if (ctx.path.match(/\.(js|gif|jpg|png|pdf|json)$/)) return false;
+      // The Accept header is a good proxy for whether SSR should happen
+      // Requesting an HTML page via the browser url bar generates a request with `text/html` in its Accept headers
+      // XHR/fetch requests do not have `text/html` in the Accept headers
+      if (!ctx.headers.accept) return false;
+      if (!ctx.headers.accept.includes('text/html')) return false;
+      return true;
+    };
+  },
+});
+export {SSRDecider};
+
+export default function createSSRPlugin({
+  element,
+  ssrDecider,
+}: {
+  element: any,
+  ssrDecider: SSRDeciderService,
+}) {
+  return async function ssrPlugin(ctx: Context, next: () => Promise<void>) {
+    if (!ssrDecider(ctx)) return next();
+
+    const template = {
+      htmlAttrs: {},
+      bodyAttrs: {},
+      title: '',
+      head: [],
+      body: [],
+    };
+    ctx.element = element;
+    ctx.rendered = '';
+    ctx.template = template;
+    ctx.type = 'text/html';
+    
+    await next();
+
+    // Allow someone to override the ssr by setting ctx.body
+    // This is especially useful for things like ctx.redirect
+    if (ctx.body && ctx.respond !== false) {
+      return;
+    }
+                    
+    ctx.body = multi_stream
```

Comment:
Do you have benchmarks for this? It feels that this is streaming only in a shallow sense of the word, and without having seen data, I suspect that this would only add overheads on top of simply flushing out an entire string at once.

Ideally, one should be able to hook into streaming APIs such as React 16's `renderToNodeStream ` so that React could push bytes out as they became available, rather than "blocking" until the entire React tree was serialized.

Reply:
@lhorie I know about `renderToNodeStream`, but i feel that function is too ambitous, not sure if it's good or not.

Reply:
This kind of `string-to-stream` is interesting, because you can cache it however you like, for example, by using a Redis cache. With `renderToNodeStream`, you lack of control over everything.

Reply:
@lhorie what's your recommended way to benchmark our code ?

Reply:
I put a benchmark here:
https://gist.github.com/revskill10/8386bbcf54c2f2d0860c98e9eaccdb06